rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+   pub async fn destroy(path: Path, object_store: Arc<dyn ObjectStore>, soft: bool) -> Result<(), SlateDbError> {
+       …
+   }
+}
+
+/// Configuration options for the database reader. These options are set on client startup.
+#[derive(Clone)]
+pub struct DbReaderOptions {
+    /// How frequently to poll for new manifest files. Refreshing the manifest file allows readers to detect newly compacted data.
+    pub manifest_poll_interval: Duration,
+
+
+    /// For readers that refresh their checkpoint, this specifies the lifetime to use for the created checkpoint. The checkpoint’s expire time will be set to the current time plus this value. If not specified, then the checkpoint will be created with no expiry, and must be manually removed.
+    pub checkpoint_lifetime: Option<Duration>,
+
+
+    /// The max size of a single in-memory table used to buffer WAL entries
+    /// Defaults to 64MB
+    pub max_memtable_bytes: u64
+}
+
+
+struct DbReader {
+    …
+}
+
+impl DbReader {
+    /// Creates a database reader that can read the contents of a database (but cannot write any
+    /// data). The caller can provide an optional checkpoint. If the checkpoint is provided, the
+    /// reader will read using the specified checkpoint and will not periodically refresh the checkpoint. Otherwise, the reader creates a new checkpoint pointing to the current manifest and refreshes it periodically as specified in the options.
+    pub async fn open(
+        path: Path,
+        object_store: Arc<dyn ObjectStore>,
+        checkpoint: Option<UUID>,
+        options: DbReaderOptions
+    ) -> Result<Self, SlateDBError> {...}
+
+    /// Read a key an return the read value, if any.  
+  pub async fn get(&self, key: &[u8]) -> Result<Option<Bytes>, SlateDBError> {...}
+}
+```
+
+Finally, we’ll also allow users to create checkpoints using the admin CLI:
+
+```bash
+$ slatedb create-checkpoint --help
+Create a new checkpoint pointing to the database's current state
+
+Usage: slatedb --path <PATH> create-checkpoint [OPTIONS]
+
+Options:
+  -l, --lifetime <LIFETIME>  Optionally specify a lifetime for the created checkpoint. You can specify the lifetime in a human-friendly format that uses years/days/min/s, e.g. "7days 30min". The checkpoint's expiry time will be set to the current wallclock time plus the specified lifetime. If the lifetime is not specified, then the checkpoint is set with no expiry and must be explicitly removed
+
+$ slatedb refresh-checkpoint --help
+Refresh an existing checkpoint's expiry time. This command will look for an existing checkpoint and update its expiry time using the specified lifetime
+
+Usage: slatedb --path <PATH> refresh-checkpoint [OPTIONS] --id <ID>
```

Comment:
Is there a tool also to list checkpoints?

Reply:
Nope - it's worth adding. FWIW there is a tool to print the current manifest which would include this info.